### DISCUSSION TITLE
I have a decode issue with GitPython

### DISCUSSION THREAD

**Discussion by 99848873** [gitpython-developers/GitPython#1374] [Q&A] [answered]
My code is ,

![image](https://user-images.githubusercontent.com/11538922/140450755-a1a8f8d8-f757-4f1f-a9ba-f4f2a6c7ed36.png)

And there is no chinese in my repo path。

And an exception was found.

```
Exception in thread Thread-22:
Traceback (most recent call last):
  File "C:\zhoujinyuWorkSpace\MyUe4\pyscript\venv\lib\site-packages\git\cmd.py", line 106, in pump_stream
    for line in stream:
UnicodeDecodeError: 'gbk' codec can't decode byte 0xa8 in position 1297: illegal multibyte sequence

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "c:\python38\lib\threading.py", line 932, in _bootstrap_inner
    self.run()
  File "c:\python38\lib\threading.py", line 870, in run
    self._target(*self._args, **self._kwargs)
  File "C:\zhoujinyuWorkSpace\MyUe4\pyscript\venv\lib\site-packages\git\cmd.py", line 119, in pump_stream
    raise CommandError([f'<{name}-pump>'] + remove_password_if_present(cmdline), ex) from ex
git.exc.CommandError: Cmd('<stderr-pump>') failed due to: UnicodeDecodeError(''gbk' codec can't decode byte 0xa8 in position 1297: illegal multibyte sequence')
  cmdline: <stderr-pump> git pull -v origin
Update C:/workspace/AClient_Bin/AClient/Content/ repo
```

**Comment by Byron** [comment, marked as answer]
Strangely enough it looks like I have seen this before.

The problem seems to be that the python interpreter default code is `gbk`, but should be `utf-8` for most purposes. It can be changed on startup I think.

It looks like [this is how you change the default encoding](https://stackoverflow.com/questions/2276200/changing-default-encoding-of-python/7892892#7892892).

**Reply by 99848873** [reply]
I tried this, but it was not working.

**Reply by Byron** [reply]
If you say how exactly it wasn't working and what was tried this might go somewhere.

**Reply by 99848873** [reply]
![image](https://user-images.githubusercontent.com/11538922/140471048-89ad15dc-f6e5-4272-9a2a-f5775c809a27.png)
My interpreter default code is utf-8

**Reply by Byron** [reply]
<img width="683" alt="Screen Shot 2021-11-05 at 3 07 48 PM" src="https://user-images.githubusercontent.com/63622/140471800-7cdade7c-117e-49d9-9ae2-b5542bf6113c.png">

What about setting the IO encoding?

**Reply by 99848873** [reply]
I had tried other function. 
![image](https://user-images.githubusercontent.com/11538922/140471741-43493d5e-ee9a-4a29-90ee-596e089c45a2.png)
Got another failed
Cmd('git') failed due to: exit code(1)
  cmdline: git pull -v Develop

**Reply by 99848873** [reply]
All are utf-8
![image](https://user-images.githubusercontent.com/11538922/140471960-5ba9f755-61b8-4d6d-b415-398d1c7c2cf3.png)

**Reply by 99848873** [reply]
I tried repo.git.pull() and without any parameters，it was wroking.
![image](https://user-images.githubusercontent.com/11538922/140475179-b974f93b-b37b-4030-9e2f-e47543d65392.png)

**Reply by Byron** [reply]
Where is the 'gbk' codec coming from? Is it some other program code outside of GitPython that sets it? What are these values after executing the code running GitPython?

**Comment by EliahKagan** [comment]
Some further information: On Windows `gbk` refers to [code page 936](https://en.wikipedia.org/wiki/Code_page_936_(Microsoft_Windows)), and it is the default ANSI active code page on Windows in some locales, especially where Chinese is spoken. Where English is spoken, [code page 1252](https://en.wikipedia.org/wiki/Windows-1252) is most often the default in Windows.

**Python can be made to use UTF-8** instead of the ANSI active code page by [running the interpreter with `-X utf8` or setting the `PYTHONUTF8` environment variable to the value `1`](https://peps.python.org/pep-0540/). I have found that this (the `-X` way) has been sufficient to overcome some limitations, such as #1747 before it was fixed. I am unsure if it is sufficient to overcome all such limitations; I have not at this point tested to see if it ameliorates #1362.

Python will also use UTF-8 when instructed to do so, by passing `encoding="utf-8"` where applicable, which is something that has to be done in the code of a program. *If* running Python with `-X utf8` is sufficient to work around most or all encoding limitations in GitPython, then it might be possible to fix them by using `encoding="utf-8"` in most or all calls to functions that recognize it (when operating with text). But I am unsure if that would work, because there is also the behavior of external programs, particularly GitPython's `git` subprocesses, to consider.

Unlike other systems where the most closely related locale settings are more readily changed and can be changed easily for specific commands using simple and well-documented techniques, the ANSI active code page on a Windows system is a system-level setting (unlike the interface language setting and preferred language settings), which can be changed, but only for the whole system, and this is rarely done. [This post](https://serverfault.com/a/836221/99345) and its linked sources provide some more information about that. Windows can also be [made to use UTF-8](https://stackoverflow.com/a/57134096) where it would otherwise use a localized code page, though this is considered "beta" and some applications might have problems.
